Foundations of the Auto-Fix Feature

mostafa opened the pull request:
This PR lays the foundation for the auto-fix functionality, enabling the tool to suggest and apply security fixes to GitHub Actions workflows, while preserving YAML formatting, structure, and comments on a best-effort basis.

### Key Features
* **Fix Framework**: introduces a foundational Fix abstraction that allows each audit finding to include one or more suggested or auto-applicable fixes. While all audit rules are designed to support fix generation, this PR includes fix capabilities only for the template-injection rule as a starting point. Fixes are applied as direct YAML patches using JSON pointers, avoiding the lossy YAML → JSON → Patch → YAML workflow that would strip comments and alter formatting.
* **YAML Patch Engine**: custom-built `yaml_patch` module to apply structured patches while retaining YAML comments and formatting.
* **CLI Options for Fixing**: adds a CLI flag to support auto-fixing:
    * `--fix`: apply available fixes automatically.
* **Testing**: adds extensive test coverage for patching and fix logic.

### How to Use
```bash
# Apply fixes with confirmation
zizmor --fix workflow.yml
```

### Notes
This is a large and foundational PR split from the massive #854 to make review easier. It's still evolving, and feedback is welcome, happy to adapt it further as needed. (This is the smallest possible denomination I could create.)

xref https://github.com/zizmorcore/zizmor/issues/28

woodruffw commented on crates/zizmor/src/main.rs:
```diff
@@ -141,6 +143,22 @@ struct App {
+    #[arg(long)]
+    fix: bool,
+
+    /// Skip confirmation prompt when applying fixes.
+    #[arg(long, requires = "fix")]
+    yes: bool,
+
+    /// Show what fixes would be applied without actually writing them.
+    #[arg(long, requires = "fix")]
+    dry_run: bool,
+
+    /// Show a diff of the changes before applying them.
+    #[arg(long, requires = "fix")]
+    diff: bool,
```
Mind deferring these to a future PR, just to keep this one small(er)? I think we can start with `--fix` and then look at working these in.

mostafa replied:
I separated the functionality into a `fix.go` file for easier review. Also removed the extra flags.

woodruffw commented on crates/zizmor/src/main.rs:
```diff
@@ -653,6 +718,241 @@ fn run() -> Result<ExitCode> {
     };
 
+    if app.fix {
+        use std::collections::HashMap;
```
Nitpick: `use` at the top of the file, please!

woodruffw commented on crates/zizmor/src/registry.rs:
```diff
@@ -110,9 +110,54 @@ impl InputKey {
+                .map_err(|e| InputError::Io(e))?
+                .try_into()
+                .map_err(|e| InputError::Other(anyhow!("failed to convert path to UTF-8: {}", e)))?
+        } else {
+            // For non-existing files, resolve relative to current directory
+            std::env::current_dir()
+                .map_err(|e| InputError::Io(e))?
+                .join(path.as_ref())
+                .try_into()
+                .map_err(|e| InputError::Other(anyhow!("failed to convert path to UTF-8: {}", e)))?
+        };
+
+        // Also canonicalize the prefix if provided, to ensure consistent path resolution
+        let canonicalized_prefix = if let Some(prefix) = prefix.as_ref() {
+            if prefix.as_ref().exists() {
+                Some(
+                    prefix
+                        .as_ref()
+                        .canonicalize()
+                        .map_err(|e| InputError::Io(e))?
+                        .try_into()
+                        .map_err(|e| {
+                            InputError::Other(anyhow!("failed to convert prefix to UTF-8: {}", e))
+                        })?,
+                )
+            } else {
+                // For non-existing prefix, resolve relative to current directory
+                Some(
+                    std::env::current_dir()
+                        .map_err(|e| InputError::Io(e))?
+                        .join(prefix.as_ref())
+                        .try_into()
+                        .map_err(|e| {
+                            InputError::Other(anyhow!("failed to convert prefix to UTF-8: {}", e))
+                        })?,
+                )
+            }
+        } else {
+            None
+        };
```
Could you provide some context for these changes? Off the top of my head, I _think_ this is going to cause problems: `given_path` is meant to be the exact path the user gives us without any canonicalization, which we do because SARIF consumers like GitHub struggle with absolute paths.

For some context, see:

* https://github.com/zizmorcore/zizmor/pull/469
* https://github.com/zizmorcore/zizmor/pull/572

(Speculating: if this change is being motivated by the `file_fixes` mapping, then I _think_ the `AuditInput` suggestion should eliminate the need for this change 🙂)

mostafa replied:
You are right. We no longer need this change.
https://github.com/zizmorcore/zizmor/pull/858/commits/72f704e6c21b87d58dcd562d417a1d3ef7f0d6b7

woodruffw commented on crates/zizmor/Cargo.toml:
```diff
@@ -63,6 +63,8 @@ tree-sitter.workspace = true
 tree-sitter-powershell.workspace = true
 yamlpath.workspace = true
+diff = "0.1.13"
+jsonptr = "0.7"
```
Nit: where does this dep get used? I tried to find it in `yaml_patch` but couldn't 🙂

mostafa replied:
Leftover from previous testings. It is now manually implemented in the code.

mostafa replied:
https://github.com/zizmorcore/zizmor/pull/858/commits/6bfb7b208210dd53dd34c6d4b5d390f44ded9db6

woodruffw commented on Cargo.toml:
```diff
@@ -48,7 +48,7 @@ reqwest = { version = "0.12.15", default-features = false }
 serde = { version = "1.0.219", features = ["derive"] }
 serde-sarif = "0.8.0"
-serde_json = "1.0.140"
+serde_json = { version = "1.0.140", features = ["preserve_order"] }
```
Not an issue, but out of curiosity: where do we need order-preserving maps here?

(Specifically I'm curious because this is for `serde_json`, not `serde_yaml`.)

mostafa replied:
Also a leftover from when I was trying the older approach of conversions between JSON and YAML and I needed to have this feature to preserve order of keys in the output document, because otherwise it was sorting it and messing it up entirely. Removed.

woodruffw commented on crates/zizmor/src/yaml_patch/mod.rs:
```diff
@@ -0,0 +1,1926 @@
+/// - "/permissions/contents" → Query for nested key
+/// - "/jobs/0/steps/1" → Query for array indices
+/// - "/path/with~1slash" → Query for key containing "/"
+pub fn parse_json_pointer_to_query(path: &str) -> Result<yamlpath::Query, YamlPatchError> {
```
Architectural question: instead of needing an additional bit of syntax (JSON pointers) for patches here, perhaps we could have patch paths/routes reuse the existing `SymbolicLocation` or `Route` APIs?

If I'm thinking about this right, I _think_ that would allow the "finding" and "fix" layers to share more state/common code here. 

(This is not a hard requirement for this PR or feature, but I'm curious if you have thoughts about it! See https://github.com/zizmorcore/zizmor/issues/779 for some more context as well 🙂)

mostafa replied:
If you already have an idea how to implement this, feel free to edit this PR.

woodruffw commented on crates/zizmor/src/yaml_patch/mod.rs:
```diff
@@ -0,0 +1,1926 @@
+    }
+}
+
+/// Check if a string needs to be quoted in YAML
```
Nitpick: I'm okay with removing this entirely and having the fix mode always perform quoting.

For context: Rust/LLVM struggle to reduce the number of linear scans needed by each `||` here; see [this Godbolt link](https://godbolt.org/z/Yse3f4ff8). An alternative would be to use a regex to accomplish the same thing, but IMO that'll be brittle/hard to maintain so I'm okay with a slightly conservative quoting approach instead 🙂

mostafa commented:
@woodruffw Thanks for the review! I'll have time to address them tomorrow.

woodruffw commented:
Thanks, I'll do that tonight!

mostafa commented on crates/zizmor/src/yaml_patch/mod.rs:
```diff
@@ -0,0 +1,1926 @@
+        assert!(!result.contains("contents: read"));
+
+        println!("Full demo test passed!");
+        println!("📝 Original YAML had comments preserved while applying transformations");
```
I replaced them with asserts wherever possible: https://github.com/zizmorcore/zizmor/pull/858/commits/480ff0504c5aa2ff7a59c4a192038e7e5eaf1e58.

mostafa commented on crates/zizmor/src/yaml_patch/mod.rs:
```diff
@@ -0,0 +1,1926 @@
+
+        println!("Step 0 extraction:");
+        println!("'{}'", doc.extract(&step0_feature));
+        println!("Byte span: {:?}", step0_feature.location.byte_span);
```
Same as above.

mostafa commented on crates/zizmor/src/main.rs:
```diff
@@ -28,12 +28,14 @@ use tracing_subscriber::{EnvFilter, layer::SubscriberExt as _, util::SubscriberI
 mod registry;
 mod state;
 mod utils;
+pub mod yaml_patch;
```
Fixed in https://github.com/zizmorcore/zizmor/pull/858/commits/9636b9980ac8f777511d32514106423b62efbdd5.

mostafa commented on crates/zizmor/src/audit/template_injection.rs:
```diff
@@ -303,14 +554,60 @@ impl Audit for TemplateInjection {
+        // Add fixes to each finding for workflow steps
+        for finding in &mut findings {
+            if let Some((script, _)) = Self::scripts_with_location(step).first() {
+                // Extract the expression from the annotation
+                if let Some(location) = finding.locations.iter().find(|l| l.symbolic.is_primary()) {
+                    if let Some(expr_start) = location.symbolic.annotation.find("may expand into") {
```
This is a very good point. I'll change the signature of the `injectable_template_expressions` function to include the context as well, so I can cleanly extract it for fixes.

mostafa replied:
Fixed in https://github.com/zizmorcore/zizmor/pull/858/commits/df4ef43aff85d89a8e8952d0e66eafba3a20341c.

mostafa commented on crates/zizmor/src/yaml_patch/mod.rs:
```diff
@@ -0,0 +1,1926 @@
+
+    let mut result = content.to_string();
+
+    if let Some(last_idx) = last_content_line_idx {
```
I thought of that as well when implementing, but was more concerned about preserving the structure of the file. I like your idea. Let me see if I can implement it.

mostafa replied:
I tried two different approaches of implementing this not only for root-level, but also for nested objects (which quickly turned out to be a hard problem) by:
1. modifying the existing yaml_patch implementation to enable addition disposition.
2. recreating the whole thing with tree-sitter-yaml to get a parse tree and work on that.

Both worked to some extent, and even though the tree-sitter version provided more flexibility, yet both have the same set of issues:
1. Line shifting is a major issue and very hard to detect with the current implementation of `Fix::apply`. At the moment, the Fixes are accumulated per Finding, and applied at the end. Almost all operations change the line count (to some extent), while the Finding still refer to the old line numbers. The other side of the coin, the content reference (yamlpath?), also might pose other set of issues.
2. Operation dependency is also another problem. Later operations see the results of earlier operations, sometimes causing unexpected behavior. This is also the cause of line shifting. We have conflicts between operations.
3. Formatting issues arise when inserting at desired positions, and not the bottom.
4. Bottom disposition is the safest, at least for now.

I am currently trying to only apply this to root-level to see if I can pull this off. Otherwise, we should probably create another issue to change this later, because it is taking too long.

mostafa replied:
I am convinced that this is a hard problem to solve. Let's revisit this at a later time, unless you have an idea how to properly solve this.

mostafa commented:
@woodruffw I am done with the changes.

mostafa commented on crates/zizmor/src/audit/template_injection.rs:
```diff
@@ -265,32 +276,292 @@ impl TemplateInjection {
+        // Convert expressions like "github.event.issue.title" to "GITHUB_EVENT_ISSUE_TITLE"
+        // For fromJSON(secrets.foo).bar patterns, extract meaningful parts
+
+        let cleaned_expr = if expr.trim().starts_with("fromJSON(secrets.") {
```
Yes. That's why I mentioned the suggestions and then fixes. For this, we can resort to suggesting how to fix it instead.

mostafa replied:
Fixed in https://github.com/zizmorcore/zizmor/pull/858/commits/88e5d846a6153e30de5b00aa7dbe5cc6a995989f.

mostafa commented on crates/zizmor/src/audit/template_injection.rs:
```diff
@@ -265,32 +276,292 @@ impl TemplateInjection {
+    }
+
+    /// Create a fix that adds explicit shell specification for workflow steps
+    fn create_shell_specification_fix(job_id: &str, step_index: usize) -> Fix {
+        Fix {
+            title: "Add explicit shell specification".to_string(),
+            description: "Add 'shell: bash' to ensure consistent variable expansion syntax across different runners. \
+                When switching to '${VARNAME}', keep in mind that different shells have different environment variable syntaxes. \
+                In particular, PowerShell (the default shell on Windows runners) uses '${env:VARNAME}'. \
+                To avoid having to specialize your handling for different runners, you can set 'shell: bash'.".to_string(),
+            apply: apply_yaml_patch!(vec![YamlPatchOperation::MergeInto {
+                path: format!("/jobs/{}/steps/{}", job_id, step_index),
+                key: "shell".to_string(),
+                value: serde_yaml::Value::String("bash".to_string()),
+            }]),
+        }
```
This makes sense! I'll defer it to #288.

mostafa commented on crates/zizmor/src/finding/mod.rs:
```diff
@@ -442,6 +442,23 @@ pub(crate) struct Determinations {
+    /// A detailed description of the fix.
+    pub description: String,
+    /// A function that, when called, applies the fix and returns the new content (or None if not applicable).
+    pub apply: Box<dyn Fn(&str) -> anyhow::Result<Option<String>> + Send + Sync>,
```
Fixed in https://github.com/zizmorcore/zizmor/pull/858/commits/86e5736d66efc2f5db6176a1cb78654fcb557a16.

mostafa commented:
@woodruffw Thanks for the reviews! 🙏 

1. Experimental release is a good way to test user experience and I am all for it.
2. Let's give it a few iterations and it'd be good to go.
3. We can mention this feature (its PR, discussions, related tickets and such) in the next release notes and mention it'll be available in the upcoming release, so as to gather feedback. This is just a suggestion. You ultimately decide what to do with it.
4. I fixed your suggestions and also updated the ticket description to match the current functionality.

mostafa commented:
@woodruffw 
One more thing! Currently the suggestions are left to the findings. I can create Fixes with no-op apply functions with only the title and description, and then enumerate them in the end with the given suggestion. However, I guess that would be superfluous and redundant.

woodruffw commented on crates/zizmor/src/fix.rs:
```diff
@@ -0,0 +1,186 @@
+            println!("{}", format!("\nFixes").green().bold());
+            let num_fixes = file_applied_fixes.len();
+            for (ident, fix, finding) in file_applied_fixes {
+                let line_info = if let Some(line) = get_primary_line_number(finding) {
```
FYI you can unwrap here, rather than handling a superfluous `None` case -- the finding builder APIs enforce the presence of a primary location, so it's safe to assume one here as well 🙂 

(Alternatively you could do the unwrap within `get_primary_line_number`, that might be a little cleaner.)

woodruffw commented:
Thanks again @mostafa! I'll take over the PR for a bit here to clean up the lints, etc.
